### PR TITLE
Breaking: make no-redeclare stricter (fixes #11370, fixes #11405)

### LINKED ISSUE #11370: Proposal: no-redeclare reports /*globals foo*/ comments if it's redeclaration
**What rule do you want to change?**

- [no-redeclare] rule

**Does this change cause the rule to produce more or fewer warnings?**

- more warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

- I want to change the default behavior. But I'm OK if it's behind an option.

**Please provide some example code that this change will affect:**

config:

```yml
globals:
  window: readable
rules:
  no-redeclare:
    - error
    - builtinGlobals: true
```

code:

```js
/*globals window */
```

**What does the rule currently do for this code?**

- Does nothing.

**What will the rule do after it's changed?**

- Reports the `window` in `/*globals*/` comment because it's declared already.

For a reference, [no-unused-vars] rule reports unused variables that `/*globals*/` comment defined. For example:

> ```js
> /*globals foo */ //ERROR: 'foo' is defined, but never used.
> ```
>
> https://eslint.org/demo/#eyJ0ZXh0IjoiLypnbG9iYWxzIGZvbyAqLyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11bnVzZWQtdmFycyI6Mn0sImVudiI6e319fQ==

Therefore, this change makes consistent between [no-unused-vars] rule and [no-redeclare] rule.

**Are you willing to submit a pull request to implement this change?**

Yes, I will do.

[no-redeclare]: https://eslint.org/docs/rules/no-redeclare
[no-unused-vars]: https://eslint.org/docs/rules/no-unused-vars

### LINKED ISSUE #11405: Proposal: change no-redeclare default behavior
**What rule do you want to change?**

- [no-redeclare]

**Does this change cause the rule to produce more or fewer warnings?**

- More Warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

- It changes the default value of `builtinGlobals` option to `true`.

**Please provide some example code that this change will affect:**

```js
/*eslint-env browser */
/*eslint no-redeclare: error */

var top = 100
```

**What does the rule currently do for this code?**

- [No errors](https://eslint.org/demo/#eyJ0ZXh0IjoiLyplc2xpbnQtZW52IGJyb3dzZXIgKi9cclxuLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBlcnJvciAqL1xyXG5cclxudmFyIHRvcCA9IDEwMCIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJicm93c2VyIjpmYWxzZX19fQ==).

**What will the rule do after it's changed?**

- [Warn the `top` variable as a redeclaration](https://eslint.org/demo/#eyJ0ZXh0IjoiLyplc2xpbnQtZW52IGJyb3dzZXIgKi9cclxuLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBbZXJyb3IsIHtidWlsdGluR2xvYmFsczogdHJ1ZX1dICovXHJcblxyXG52YXIgdG9wID0gMTAwIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7ImJyb3dzZXIiOmZhbHNlfX19) because `browser` environment has it.

**Are you willing to submit a pull request to implement this change?**

Yes.

**Context:**

- We cannot redeclare the `top` variable, browsers ignore the `var top` declaration silently. This can cause confusion and I believe that it's worth to warn by default.
- Originate `no-undef` rule had warned redeclarations of globals by default, but it was removed in https://github.com/eslint/eslint/issues/4504. I think that we can re-enable it on `no-redeclare` rule.

[no-redeclare]: https://eslint.org/docs/rules/no-redeclare

### PR DESCRIPTION
**Pull request by mysticatea**
**What is the purpose of this pull request? (put an "X" next to item)**

[X] Changes an existing rule
[X] Add something to the core

- fixes #11370 
- fixes #11405

**What changes did you make? (Give an overview)**

This PR makes `no-redeclare` rule stricter.

- Implements https://github.com/eslint/rfcs/pull/17
- Adds `/*globals*/` comment check to `no-redeclare` rule. (#11370)
- Enables `{ builtinGlobals: true }` option by default. (#11405)
- ~~Adds some items to the migration guide.~~ (moved to #11515)

And some chore:

- Removes the accesses to `parserOptions` from `no-redeclare` rule. The `parserOptions` is not safe because each parser has different options.
- Moves a utility that determines the location of a variable name in `/*globals*/` comments to `ast-utils` because it's shared with `no-unused-vars` rule and `no-redeclare` rule.
- Adds several tests for lexical bindings into `no-redeclare` rule along with a loose parser that doesn't throw syntax errors on redeclaration of lexical bindings.
- ~~Makes `eslint-plugin/test-case-property-ordering` rule's option more reasonable.~~ (moved to #11511)

**Is there anything you'd like reviewers to focus on?**

- Should `no-redeclare` rule handle lexical bindings still? Since acorn 5.0.0, it throws syntax errors by such a redeclaration.

### CODE REVIEW

File: docs/user-guide/migrating-to-6.0.0.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Migrating to v6.0.0
+
+ESLint v6.0.0 is the fifth major version release. We have made a few breaking changes in this release, but we expect that most users will be able to upgrade without any modifications to their build. This guide is intended to walk you through the breaking changes.
+
+The lists below are ordered roughly by the number of users each change is expected to affect, where the first items are expected to affect the most users.
+
+### Breaking changes for users
+
+- [`no-redeclare` rule now checks the redeclarations by `/* globals */` directive comments](#no-redeclare-and-comments)
+- [`no-redeclare` rule now checks the redeclarations with built-in globals by default](#no-redeclare-and-builtins)
```
Comment (not-an-aardvark) [inline docs/user-guide/migrating-to-6.0.0.md]: To avoid merge conflicts, I was thinking it might be better to create the migration guide separately rather than adding an entry in each breaking commit.

File: lib/rules/no-redeclare.js [hunk 2]
```diff
@@ -33,72 +45,128 @@ module.exports = {
 
     create(context) {
         const options = {
-            builtinGlobals: context.options[0] && context.options[0].builtinGlobals
+            builtinGlobals: Boolean(
+                context.options.length === 0 ||
+                context.options[0].builtinGlobals !== false
```
Comment (not-an-aardvark) [inline lib/rules/no-redeclare.js]:
```suggestion
                context.options[0].builtinGlobals
```

Since `default` is used, I think this will always be a boolean.

File: lib/rules/no-unused-vars.js [hunk 3]
```diff
@@ -632,21 +598,26 @@ module.exports = {
...
                         context.report({
                             node: unusedVar.identifiers[0],
                             message: unusedVar.references.some(ref => ref.isWrite())
                                 ? getAssignedMessage()
                                 : getDefinedMessage(unusedVar),
                             data: unusedVar
                         });
+
+                    // If there are no regular declaration, report the first `/*globals*/` comment directive.
+                    } else if (unusedVar.eslintExplicitGlobalComments && unusedVar.eslintExplicitGlobalComments.length > 0) {
```
Comment (not-an-aardvark) [inline lib/rules/no-unused-vars.js]: To clarify how this works, will `eslintExplicitGlobalComments` always be nonempty if the property exists?

File: package.json [hunk 4]
```diff
@@ -123,6 +124,6 @@
   ],
   "license": "MIT",
   "engines": {
-    "node": "^6.14.0 || ^8.10.0 || >=9.10.0"
+    "node": "^8.10.0 || ^10.13.0 || >=11.10.1"
```
Comment (not-an-aardvark) [inline package.json]: Can we make this change in a separate PR?

File: .eslintrc.js [hunk 5]
```diff
@@ -17,7 +17,22 @@ module.exports = {
...
+                "code",
+                "output",
+                "options",
+                "parser",
+                "parserOptions",
+                "globals",
+                "env",
+                "errors"
+            ]
+        ],
```
Comment (not-an-aardvark) [inline .eslintrc.js]: Can we make this change in a separate PR? (To avoid merge conflicts later, we could merge it as a `Chore` soon, before v6 prereleases start.)

File: docs/developer-guide/working-with-rules.md [hunk 6]
```diff
@@ -169,6 +169,13 @@ This method returns the scope which has the following types:
 **※2** Only if the `for` statement defines the iteration variable as a block-scoped variable (E.g., `for (let i = 0;;) {}`).<br>
 **※3** The scope of the closest ancestor node which has own scope. If the closest ancestor node has multiple scopes then it chooses the innermost scope (E.g., the `Program` node has a `global` scope and a `module` scope if `Program#sourceType` is `"module"`. The innermost scope is the `module` scope.).
 
+The returned value is a `Scope` object that [`eslint-scope` package defined](scope-manager-interface.md). The `Variable` objects of global variables have some additional properties.
+
+* `variable.writeable` (`boolean`) ... If `true`, this global variable can be assigned arbitrary value. Otherwise, this global variable is read-only.
+* `variable.eslintExplicitGlobal` (`boolean`) ... If `true`, this global variable was defined by a `/* globals */` directive comment in the source code file.
+* `variable.eslintExplicitGlobalComments` (`Comment[]`) ... The array of `/* globals */` directive comments which defined this global variable in the source code file.
```
Comment (not-an-aardvark) [inline docs/developer-guide/working-with-rules.md]: Will this property always exist, or is it only defined on variables that have at least one comment?

File: package.json (hunk 4 above)
Comment (mysticatea) [reply package.json]: Yeah, I noted it in the PR description:

> Changes engines.node to use Object.entries for now (this is the reason that this PR has "don't merge" label. I assume this PR will be merged after the PR which updates engines.node was merged.)

File: docs/user-guide/migrating-to-6.0.0.md (hunk 1 above)
Comment (mysticatea) [reply docs/user-guide/migrating-to-6.0.0.md]: I got it. I will remove the migration guide.

File: package.json (hunk 4 above)
Comment (not-an-aardvark) [reply package.json]: Oh, sorry, I missed that. Ok, sounds fine to me.

File: lib/rules/no-unused-vars.js (hunk 3 above)
Comment (mysticatea) [reply lib/rules/no-unused-vars.js]: You are right. Indeed, this comparison is redundant.

File: docs/developer-guide/working-with-rules.md (hunk 6 above)
Comment (mysticatea) [reply docs/developer-guide/working-with-rules.md]: It's only defined on variables that have at least one comment. I will add the description.

File: .eslintrc.js (hunk 5 above)
Comment (mysticatea) [reply .eslintrc.js]: Done #11511

File: docs/developer-guide/working-with-rules.md [hunk 7]
```diff
@@ -169,6 +169,13 @@ This method returns the scope which has the following types:
 **※2** Only if the `for` statement defines the iteration variable as a block-scoped variable (E.g., `for (let i = 0;;) {}`).<br>
 **※3** The scope of the closest ancestor node which has own scope. If the closest ancestor node has multiple scopes then it chooses the innermost scope (E.g., the `Program` node has a `global` scope and a `module` scope if `Program#sourceType` is `"module"`. The innermost scope is the `module` scope.).
 
+The returned value is a `Scope` object that [`eslint-scope` package defined](scope-manager-interface.md). The `Variable` objects of global variables have some additional properties.
```
Comment (platinumazure) [inline docs/developer-guide/working-with-rules.md]: Possible wording improvement:

```suggestion
The returned value is a [`Scope` object](scope-manager-interface.md) defined by the `eslint-scope` package. The `Variable` objects of global variables have some additional properties.
```

File: lib/util/ast-utils.js [hunk 8]
```diff
@@ -1342,5 +1343,29 @@ module.exports = {
...
+        // To ignore the first text "global".
+        namePattern.lastIndex = comment.value.indexOf("global") + 6;
+
+        // Search a given variable name.
+        const match = namePattern.exec(comment.value);
+
+        // Convert the index to loc.
+        return sourceCode.getLocFromIndex(
+            comment.range[0] +
+            2 + // 2 is "/*".length
```
Comment (platinumazure) [inline lib/util/ast-utils.js]: Might as well be explicit and a little more future-proof, no?

```suggestion
            "/*".length +
```

File: tests/lib/rules/no-redeclare.js [hunk 9]
```diff
@@ -98,13 +124,203 @@ ruleTester.run("no-redeclare", rule, {
...
+            options: [{ builtinGlobals: true }],
+            globals: { a: "writable" },
+            errors: [
+                { message: "'a' is already defined as a built-in global variable.", type: "Block" }
+            ]
+        },
+        {
+            code: "/*globals a */ /*globals a */",
+            errors: [
+                { message: "'a' is already defined.", type: "Block" }
```
Comment (platinumazure) [inline tests/lib/rules/no-redeclare.js]: Could you add line/column to show which comment is being flagged?

**Comment by mysticatea** [comment]
Thank you very much!
